tests/integration/exec.bats

```diff
@@ -167,3 +167,115 @@ function check_exec_debug() {
 	[[ "${output}" == *"level=debug"* ]]
 	check_exec_debug "$output"
 }
+
+@test "runc exec --cgroup sub-cgroups [v1]" {
+	requires root cgroups_v1 cgroupns
+
+	set_cgroups_path
+	set_cgroup_mount_writable
+	# Enable CGROUPNS (to simplify test case working with cgroup paths).
+	update_config '.linux.namespaces += [{"type": "cgroup"}]'
+
+	__runc run -d --console-socket "$CONSOLE_SOCKET" test_busybox
+	testcontainer test_busybox running
+
+	# Check we can't join non-existing subcgroup.
+	runc exec --cgroup nonexistent test_busybox cat /proc/self/cgroup
+	[ "$status" -ne 0 ]
+	[[ "$output" == *" adding pid "*"/nonexistent/cgroup.procs: no such file "* ]]
+
+	# Check we can't join non-existing subcgroup (for a particular controller).
+	runc exec --cgroup cpu:nonexistent test_busybox cat /proc/self/cgroup
+	[ "$status" -ne 0 ]
+	[[ "$output" == *" adding pid "*"/nonexistent/cgroup.procs: no such file "* ]]
+
+	# Check we can't specify non-existant controller.
+	runc exec --cgroup whaaat:/ test_busybox true
+	[ "$status" -ne 0 ]
+	[[ "$output" == *"unknown controller "* ]]
+
+	# Check we can join top-level cgroup (implicit).
+	runc exec test_busybox sh -euc "! grep -v ':/$' /proc/self/cgroup"
+	[ "$status" -eq 0 ]
+	[ "$output" = "" ]
+
+	# Check we can join top-level cgroup (explicit).
+	runc exec --cgroup / test_busybox sh -euc "! grep -v ':/$' /proc/self/cgroup"
+	[ "$status" -eq 0 ]
+	[ "$output" = "" ]
+
+	# Create a few subcgroups.
+	runc exec test_busybox sh -euc "mkdir -p /sys/fs/cgroup/memory/submem /sys/fs/cgroup/cpu/subcpu /sys/fs/cgroup/cpuacct/subcpu"
+	[ "$status" -eq 0 ]
+
+	# Check that explicit --cgroup works.
+	runc exec --cgroup memory:submem --cgroup cpu,cpuacct:subcpu test_busybox cat /proc/self/cgroup
+	[ "$status" -eq 0 ]
+	[[ "$output" == *":memory:/submem"* ]]
+	[[ "$output" == *":cpu"*":/subcpu"* ]]
+}
+
+@test "runc exec --cgroup subcgroup [v2]" {
```

Comment:
Can we have a test and doc for host cgroupns mode?

Reply:
It should be all the same (IOW nothing special about it). As noted in the comment above, the cgroupns is enabled only to simplify the test code (as we don't have to figure out in-container paths to sub-cgroups).

I can certainly rewrite the test to not be dependent on cgroupns. Let me try that.

Reply:
Made it working without cgroupns; PTAL @AkihiroSuda.

The major differences are:
1. Instead of looking for `/` in /proc/self/cgroup, we look for `$REL_CGROUPS_PATH`.
2. The rdma controller has to be filtered out of /proc/cgroup/path in two cases, as we do not support for it yet (see #2883). It works as is with cgroupns since if enabled, /proc/<pid>cgroup does not (and can not) distinguish between a process being in host root cgroup and container root cgroup -- showing `/` as the path to the cgroup in both cases.

IOW it works the same as without cgroupns.

Reply:
Now that #2883 is merged, I removed filtering out rdma.

```diff
diff --git a/tests/integration/exec.bats b/tests/integration/exec.bats
index c605aec2..24444d1e 100644
--- a/tests/integration/exec.bats
+++ b/tests/integration/exec.bats
@@ -213,14 +213,12 @@ function check_exec_debug() {
        [[ "$output" == *"unknown controller "* ]]
 
        # Check we can join top-level cgroup (implicit).
-       # XXX: rdma is not yet supported so we filter it out.
-       runc exec test_busybox sh -euc "grep -vF :rdma: /proc/self/cgroup"
+       runc exec test_busybox cat /proc/self/cgroup
        [ "$status" -eq 0 ]
        ! grep -v ":$REL_CGROUPS_PATH\$" <<<"$output"
 
        # Check we can join top-level cgroup (explicit).
-       # XXX: rdma is not yet supported so we filter it out.
-       runc exec --cgroup / test_busybox sh -euc "grep -vF :rdma: /proc/self/cgroup"
+       runc exec --cgroup / test_busybox cat /proc/self/cgroup
        [ "$status" -eq 0 ]
        ! grep -v ":$REL_CGROUPS_PATH\$" <<<"$output"
 
```

Strictly speaking, in this test we have to only check the controllers that we know about, but I'm too lazy to implement it right now (let's wait for a new controller to appear first -- maybe we will add it faster if CI will fail).